### PR TITLE
Improve client updater UX

### PR DESCRIPTION
**Pull request by duk3luk3**
* Never show same version client already has as update
* Don't call old versions "updates"
* Don't show old versions by default

### CODE REVIEW

File: src/client/updater.py [hunk 1]
```diff
@@ -22,23 +22,28 @@ class UpdateSettings:
...
+
+        server_releases = [release for release in releases if release['branch']=='server']
+        stable_releases = [release for release in releases if release['branch']=='stable']
+        pre_releases = [release for release in releases if release['branch']=='pre']
+        beta_releases = [release for release in releases if release['branch']=='beta']
 
-            notify_stable = have_stable and self.updater_branch == UpdateBranch.Stable.name and Version(releases['stable']['new_version']) > current_version
-            notify_pre = have_pre and self.updater_branch == UpdateBranch.Prerelease.name and Version(releases['pre']['new_version']) > current_version
-            notify_beta = have_beta and self.updater_branch == UpdateBranch.Unstable.name and Version(releases['beta']['new_version']) > current_version
+        have_server = len(server_releases) > 0
```
Comment (Wesmania) [inline src/client/updater.py:33]: We can as well rely on the fact that empty lists evaluate to false and do 'if server_releases'.

File: src/client/updater.py [hunk 2]
```diff
@@ -22,23 +22,28 @@ class UpdateSettings:
...
-            return have_server or notify_stable or notify_pre or notify_beta
+        current_version = Version(config.VERSION)
+        # null out build because we don't care about it
+        current_version.build = ()
+
+        notify_stable = have_stable and self.updater_branch == UpdateBranch.Stable.name and Version(stable_releases[0]['new_version']) > current_version
+        notify_pre = have_pre and self.updater_branch == UpdateBranch.Prerelease.name and Version(pre_releases[0]['new_version']) > current_version
+        notify_beta = have_beta and self.updater_branch == UpdateBranch.Unstable.name and Version(beta_releases[0]['new_version']) > current_version
+
+        return have_server or notify_stable or notify_pre or notify_beta
```
Comment (Wesmania) [inline src/client/updater.py:46]: I think we can rearrange that to use an if / if else / else with self.updater_branch (maybe add a 'version we're interested in' variable to make things easier). Also, are we supposed to use that 'downgrade' config variable here?

File: src/client/updater.py [hunk 3]
```diff
@@ -71,30 +76,39 @@ def reset_controls(self):
...
         else:
-            self.lblUpdatesFound.setText('Client updates were found.')
+            self.lblUpdatesFound.setText('Client releases were found.')
 
         if len(self.releases) > 0:
             currIdx = 0
-            preferIdx = 0
+            preferIdx = -1
 
+            labels = dict([('server', 'Server Version'), ('stable', 'Stable Version'), ('pre', 'Stable Prerelease'), ('beta', 'Unstable')])
```
Comment (Wesmania) [inline src/client/updater.py]: Not the prettiest notation IMO

File: src/client/updater.py [hunk 4]
```diff
@@ -71,30 +76,39 @@ def reset_controls(self):
...
+        if any([release['branch']=='server' for release in self.releases]):
             self.lblUpdatesFound.setText('Your client version is outdated - you must update to play.')
         else:
-            self.lblUpdatesFound.setText('Client updates were found.')
+            self.lblUpdatesFound.setText('Client releases were found.')
 
         if len(self.releases) > 0:
             currIdx = 0
-            preferIdx = 0
+            preferIdx = -1
```
Comment (Wesmania) [inline src/client/updater.py]: None would be nice too (just a personal preference / nitpick).

File: src/client/updater.py [hunk 5]
```diff
@@ -71,30 +76,39 @@ def reset_controls(self):
...
-                        preferIdx = currIdx
+                is_update = ' [New!]' if Version(release['new_version']) > current_version else ''
 
-                    currIdx = currIdx + 1
+                self.cbReleases.insertItem(99, '{} {}{}'.format(label, release['new_version'], is_update), release)
+
+                if prefer and preferIdx == -1:
+                    preferIdx = currIdx
+
+                currIdx = currIdx + 1
```
Comment (Wesmania) [inline src/client/updater.py]: We can just iterate over enumerate(self.releases) instead.

File: src/client/updater.py [hunk 6]
```diff
@@ -141,13 +155,9 @@ def showSettings(self):
 @with_logger
 class UpdateChecker(QObject):
     gh_releases_url = Settings.persisted_property('updater/gh_release_url', type=str, default_value='https://api.github.com/repos/FAForever/client/releases?per_page=20')
+    updater_downgrade = Settings.persisted_property('updater/downgrade', type=bool, default_value=False)
```
Comment (Wesmania) [inline src/client/updater.py:156]: Why is it in 2 places at once? Is persisted_property capable of handling such a case (specifically does it update the other value when one of them changes)?

File: src/client/updater.py [hunk 7]
```diff
@@ -172,11 +182,34 @@ def server_session(self):
...
                     download_url = asset['browser_download_url']
                     tag = release_dict['tag_name']
+                    v = Version(tag)
+                    # We never want to return the current version itself,
+                    # but if `updater_downgrade` is set, we do return
+                    # older releases.
+                    # strange comparison logic is because of semantic_version
+                    # so that build info is ignored
+                    if self.updater_downgrade:
+                        if not v < cv or cv < v
```
Comment (Wesmania) [inline src/client/updater.py]: Longer variable names would help readability here IMO, something like 'current_version' and 'remote_version'. And I think operator precedence screwed us over here.

**Review by Wesmania — CHANGES_REQUESTED** [review]
Some nitpicks, one logic issue.

File: src/client/updater.py (hunk 1 above)
Comment (duk3luk3) [reply src/client/updater.py:33]: I prefer not to rely on falsyness.

File: src/client/updater.py (hunk 2 above)
Comment (duk3luk3) [reply src/client/updater.py:46]: * I don't see how to make the logic more concise without making it harder to understand. I don't think turning it into a cascade of if statements would help at all.
* No, this logic is meant for deciding whether we want to show the update dialog when the client starts.

File: src/client/updater.py (hunk 6 above)
Comment (duk3luk3) [reply src/client/updater.py:156]: persisted_property is just a wrapper around Settings.get and Settings.set so yes, this is fine.

File: src/client/updater.py (hunk 7 above)
Comment (duk3luk3) [reply src/client/updater.py]: what screwed us here is that I left off the ':'. The precedence [is fine](https://docs.python.org/3/reference/expressions.html#operator-precedence).

File: src/client/updater.py (hunk 1 above)
Comment (Wesmania) [reply src/client/updater.py:33]: It's fairly pythonic IMO, but your call.

File: src/client/updater.py (hunk 2 above)
Comment (Wesmania) [reply src/client/updater.py:46]: We can change notify_* bools to check whether a given branch has a newer version (remove  the self.updater_branch check), then use self.updater_branch separately.

File: src/client/updater.py [hunk 8]
```diff
@@ -172,11 +180,34 @@ def server_session(self):
...
+                    # We never want to return the current version itself,
+                    # but if `updater_downgrade` is set, we do return
+                    # older releases.
+                    # strange comparison logic is because of semantic_version
+                    # so that build info is ignored
+                    if self.updater_downgrade:
+                        if not (release_version < client_version or client_version < release_version):
+                            return None
+                    else:
+                        if not (release_version > client_release_versionersion):
```
Comment (Wesmania) [inline src/client/updater.py]: Typo?

**Pull request merged**